File: app/src/main/java/se/curtrune/lucy/viewmodel/EditableListViewModel.java
package se.curtrune.lucy.viewmodel;


import static se.curtrune.lucy.util.Logger.log;

import android.content.Context;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.List;

import se.curtrune.lucy.classes.item.Item;
import se.curtrune.lucy.persist.ItemsWorker;


public class EditableListViewModel extends ViewModel {
    private Item parent;
    private final MutableLiveData<List<Item>> mutableItems = new MutableLiveData<>();
    public void setParent(Item parent){
        this.parent = parent;
    }
    public void init(Item parent){
        log("EditableListViewModel.init(Item)", parent.getHeading());
        this.parent = parent;
        mutableItems.setValue(new ArrayList<>());
        mutableItems.getValue().add(createChild(parent));
        log("number of items", mutableItems.getValue().size());
    }
    public void addEmptyItem(int index) {
        log("...addEmptyItem(int)", index);
        mutableItems.getValue().add(index, createChild(parent));
    }

    public LiveData<List<Item>> getItems(){
        log("EditableListViewModel.getItems()");
        return mutableItems;
    }

    public void saveAll(Context context){
        log("...saveAll()");
        List<Item> items = mutableItems.getValue();
        for(Item item: items){
            if( !item.getHeading().isEmpty()){
                ItemsWorker.insertChild(parent, item, context);
            }
        }
    }


    public void setHeading(String heading, int index, Context context){
        log("...setHeading(String, int, Context)", heading);
        mutableItems.getValue().get(index).setHeading(heading);
    }
    private Item createChild(Item parent){
        log("...createChild(Item)");
        Item child = new Item();
        child.setParent(parent);
        child.setCategory(parent.getCategory());
        return child;
    }
}

File: app/src/main/java/se/curtrune/lucy/screens/editable_list/EditableListViewModel.kt
package se.curtrune.lucy.screens.editable_list

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import se.curtrune.lucy.modules.LucindaApplication
import se.curtrune.lucy.classes.item.Item
import se.curtrune.lucy.util.Logger


class EditableListViewModel : ViewModel() {
    private val repository = LucindaApplication.appModule.repository
    private var parent: Item? = null
    private val mutableItems = MutableLiveData<MutableList<Item>>()
    fun setParent(parent: Item?) {
        this.parent = parent
    }

    fun init(parent: Item) {
        Logger.log("EditableListViewModel.init(Item)", parent.heading)
        this.parent = parent
        mutableItems.value = ArrayList()
        mutableItems.value!!.add(createChild(parent))
        Logger.log("number of items", mutableItems.value!!.size)
    }

    fun addEmptyItem(index: Int) {
        Logger.log("...addEmptyItem(int)", index)
        mutableItems.value!!.add(index, createChild(parent!!))
    }

    val items: LiveData<MutableList<Item>>
        get() {
            Logger.log("EditableListViewModel.getItems()")
            return mutableItems
        }

    fun saveAll() {
        Logger.log("...saveAll()")
        val items: List<Item>? = mutableItems.value
        for (item in items!!) {
            if (!item.heading.isEmpty()) {
                parent?.let { repository.insertChild(it, item) }
            }
        }
    }


    fun setHeading(heading: String?, index: Int, context: Context?) {
        Logger.log("...setHeading(String, int, Context)", heading)
        mutableItems.value!![index].heading = heading
    }

    private fun createChild(parent: Item): Item {
        Logger.log("...createChild(Item)")
        val child = Item()
        child.parent = parent
        child.category = parent.category
        return child
    }
}
